### PR TITLE
Implement 'max' value send

### PR DESCRIPTION
**Pull request by cliik**
This PR adds a special value 'max', which can be accepted as a spend amount. If this amount is specified, the wallet will calculate the maximum spendable amount (less fees) and send that amount. This is effectively a sweep.

PR is currently in draft state, as I need to test and think through edge cases some more.

Resolves https://github.com/mimblewimble/grin-wallet/issues/601

### CODE REVIEW

File: controller/src/command.rs [hunk 1]
```diff
@@ -350,6 +352,16 @@ where
 	C: NodeClient + 'static,
 	K: keychain::Keychain + 'static,
 {
+	// If 'max' spend, calculate the 'amount' as the maximum spendable wallet balance, less fees.
+	if args.max && args.amount == 0 {
+		let utxos = owner_api.retrieve_outputs(None, false, false, None)?.1;
+		let spendable = utxos.iter().filter_map(|c| match c.output.status {
```
Comment (phyro) [inline controller/src/command.rs:358]: here's a way to find outputs eligible to spend https://github.com/mimblewimble/grin-wallet/blob/master/libwallet/src/internal/selection.rs#L563-L570
Note that it also takes into account the 'account'. This is something I haven't thought about before. I'm not sure what should happen if you set `max` and also a specific wallet account. I'm guessing the user would expect to use max from this account? on the other hand, if the account wasn't specified, it defaults to the main account which might make it unclear that it would only use the max from the main account.

Comment (cliik) [reply controller/src/command.rs:358]: Perfect! I wondered if a method already existed for this.Thanks!

> I'm not sure what should happen if you set max and also a specific wallet account.

I think all spend commands are account specific, so IMO this one should be too. If the user specifies an account, this _should_ (it doesn't atm) spend max from that account. If the user doesn't specify an account, it should spend max from the default account. I will use the improved method you suggest, and be sure to use the account specified by the user (or the default if unspecified).

I realize I also need to add documentation for this & update the help dialog for the send command. I'll be sure to be clear about this there.

Thanks for the quick feedback!

**Comment by cliik** [comment]
I've taken the time to test the following scenarios, and all worked as expected:
* normal send/receive functionality -- still works as before :heavy_check_mark:
* send max with zero coins in wallet -- failed gracefully with insufficient funds error :heavy_check_mark:
* send max with only unspendable coins in wallet -- failed gracefully with insufficient funds error :heavy_check_mark:
* send max with mix of spendable/unspendable -- correctly computed fee & spent available coins :heavy_check_mark:
* send max with only spendable coins -- correctly computed fee & spent entire balance with no change :heavy_check_mark:

I cannot think of any other cases to test and I feel comfortable marking this ready for review. cc @yeastplume

**Comment by cliik** [comment]
@quentinlesceller the windows tests failed for a reason that appears to be unrelated to this feature:
```
thread 'slatepack_api' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: 32, kind: Uncategorized, message: "The process cannot access the file because it is being used by another process." }', controller\tests\common\mod.rs:79:50
```
Can you trigger the workflow to run again, to see if it was a transient failure?

**Comment by cliik** [comment]
> Can you trigger the workflow to run again, to see if it was a transient failure?

I decided to force-push to trigger a rerun

**Comment by davidtavarez** [comment]
I was looking for this feature recently! @cliik would you mind to write some tests for this?

**Comment by yeastplume** [comment]
The implementation looks good and seems to work well. The wallet has extensive tests to ensure as much is covered as possible and it keeps working well, so the next step would be to automate all of the tests you've outlined above. Here's the general process:

* Start an integration test by copying one of the existing test files in `controller\src\tests`. These simulate wallet-to-wallet and wallet-to-node communication so you should be able to add tests to cover all of the new cases you outlined above and check the results.
* Check the API tests to make sure the samples don't need adjustment. In this case, `send_args` aren't really used in the doctest examples, so I don't think there's much further to do here as long as the `InitTxArgs` field is documented.

When adding a feature, I usually start by developing one of the tests in the `controller\src\tests` director to exercise it, which also saves a lot of development time cause I don't need to manually run wallets and check the results. Also easy to add new test cases there to catch explicit conditions if a bug or issue comes up.

**Comment by cliik** [comment]
@davidtavarez @yeastplume I've spent several hours trying to get automated tests working for the CLI, but hitting a roadblock. Can you point me to any tests that exercise the wallet commands, instead of the wallet API? So far all I see are tests for the wallet API, which is not useful for this feature, because the change does not exist in the API methods.

Alternatively, I could re-implement this feature in the API, instead of in the CLI command processing. I didn't want to make an API change, because this feature felt like something only intended for CLI users. API users can easily calculate the appropriate send amount, and I didn't want to suggest an API change where it isn't warranted... although perhaps this was the wrong choice.

**Comment by yeastplume** [comment]
I'm of two minds as well as to whether this should be in the API, I think it may be fine to leave it as is for now and we can revisit later. If it's just adding an argument to the send args it shouldn't break any existing APIs, but it would still need more thought and consideration than placing it where you have it now.

For testing at the CLI level, look in the top level `src\tests\cmd_line_basic.rs` for examples

**Comment by davidtavarez** [comment]
@cliik @yeastplume what do you of instead of adding a parameter the API can receive `amount` as `null` and that will means to use `max`; that's how others API are doing this. Just an idea.

**Comment by cliik** [comment]
@yeastplume ahhthanks! Idk why it was so hard for me to find those CLI tests. I'll take a look later/tomorrow.

@davidtavarez in general I prefer to be explicit. If someone accidentally omits/nulls an argunent, the wallet shouldn't proceed to empty their coins.

If I was to move this to the API (which I'm starting to become in favor of), it might make more sense to change InitTxArgs.amount to an enum with two variants. One variant being the usual specified amount, and another variant being the max spend option. This would be clean & simple, and also opens the door for other variants in the future (i.e. if we wanted a variant to spend X% balance or something like that)... I dont think we should feature creep here, lol, but future extensibility is an additional benefit worth mentioning.

**Comment by cliik** [comment]
Ok, I've thought about this feature some more and I'm not happy with my original approach. After brainstorming a few better solutions, I also don't like using an enum with special variants for the amount. That would have rippling consequences in several other parts of the codebase, and the changes would be messy at best.

I'm going to close this PR for now, and discuss the new approach I have in mind in the original issue:
https://github.com/mimblewimble/grin-wallet/issues/601

**Comment by cliik** [comment]
This PR has been replaced by a better approach (IMO): https://github.com/mimblewimble/grin-wallet/pull/657
